File: 448D.cpp
#include <iostream>
#define ll unsigned long long
#define min(a,b) ((a < b) ? a : b)

using namespace std;

ll n, m, k;

inline ll check(ll num) {
  register ll ways = 0;
  for (int i = 1; i <= n; i++) ways += min(num/i, m);
  return ways;
}

int main() {
  ios::sync_with_stdio(false);
  cin >> n >> m >> k;
  register ll l = n*m, r = 0, mid;
  while(l > r) {
    mid = (l+r)/2;
    if (check(mid) < k) r = mid+1;
    else l = mid;
  }
  cout << l << endl;
  return 0;
}

File: 448D.py
n, m, k = map(int, raw_input().split())

def check(num):
  global k
  ways = 0
  for i in xrange(1, n+1):
    ways += min(m, num/i)
  return ways

l, r = n*m+1, 0
while l > r:
  mid = (r+l)/2
  if check(mid) < k:
    r = mid+1
  else:
    l = mid
print l
